net/url/url.h

```diff
@@ -26,65 +26,68 @@ SOFTWARE.
 #define BRICKS_NET_URL_URL_H
 
 #include <algorithm>
+#include <cctype>
 #include <cstring>
-#include <string>
+#include <map>
 #include <sstream>
+#include <string>
 #include <vector>
 
 #include "../../exception.h"
+#include "../../strings/printf.h"
+#include "../../strings/split.h"
 
 namespace bricks {
 namespace net {
 namespace url {
 
 struct EmptyURLException : Exception {};
-struct EmptyURLHostException : Exception {};
+// struct EmptyURLHostException : Exception {};
 
 // URL manages the mapping between the string and parsed representations of the URL. It manages:
 //
-// * host     (string)
-// * path     (string, defaults to "/", never empty).
-// * protocol (defaults to "http", empty only if set explicitly in constructor).
-// * port (defaults to the default port for supported protocols).
+// * host    (string)
+// * path    (string, defaults to "/", never empty).
+// * scheme  (defaults to "http", empty only if set explicitly in constructor).
+// * port    (defaults to the default port for supported schemes).
 //
-// When handling redirects, the previous URL can be provided to properly handle host/port/protocol.
+// When handling redirects, the previous URL can be provided to properly handle host/port/scheme.
+
+namespace impl {
 
 namespace {
-const char* const kDefaultProtocol = "http";
+const char* const kDefaultScheme = "http";
 }
 
-struct URL {
+struct URLWithoutParameters {
   std::string host = "";
   std::string path = "/";
-  std::string protocol = kDefaultProtocol;
+  std::string scheme = kDefaultScheme;
   int port = 0;
 
-  URL() = default;
+  URLWithoutParameters() = default;
 
   // Extra parameters for previous host and port are provided in the constructor to handle redirects.
-  URL(const std::string& url,
-      const std::string& previous_protocol = kDefaultProtocol,
-      const std::string& previous_host = "",
-      const int previous_port = 0) {
+  URLWithoutParameters(const std::string& url,
+                       const std::string& previous_scheme = kDefaultScheme,
+                       const std::string& previous_host = "",
+                       const int previous_port = 0) {
     if (url.empty()) {
       throw EmptyURLException();
     }
-    protocol = "";
-    size_t offset_past_protocol = 0;
+    scheme = "";
+    size_t offset_past_scheme = 0;
     const size_t i = url.find("://");
     if (i != std::string::npos) {
-      protocol = url.substr(0, i);
-      offset_past_protocol = i + 3;
+      scheme = url.substr(0, i);
+      offset_past_scheme = i + 3;
     }
 
-    const size_t colon = url.find(':', offset_past_protocol);
-    const size_t slash = url.find('/', offset_past_protocol);
-    host = url.substr(offset_past_protocol, std::min(colon, slash) - offset_past_protocol);
+    const size_t colon = url.find(':', offset_past_scheme);
```

Comment:
This colon could match the username-password separator (`http://user:pass@host:80/`).
You should likely first parse/ignore the authorization part if it is there.